Allow "enum" type

skeets23 opened the pull request:
Instead of just mapping enum type fields to act as "string", I created a custom type. This way a MySQL field like this:
```
size ENUM('x-small', 'small', 'medium', 'large', 'x-large')
```

will produce a factory like this:

```
...
'size' => $faker->randomElement(['x-small','small','medium','large','x-large']),
...
```

jasonmccreary commented:
Thanks for this.

Before it can be merged, can you match the PSR-2 coding style. Also, the Enum class seems unnecessary. Please inline it to keep everything within the current class.

skeets23 commented:
Ok, I've fixed the indenting and replaced FCN with imports. (I'm assuming that's what you were referring to?)

Unfortunately, the Enum class can't be removed. MySql `enum` fields are not supported by the `DBAL` library at all, and so to allow them, a custom data type must be used. (that I've named `customEnum` here, because `DBAL` wouldn't accept even a custom data type named `enum`).

I followed the official documentation to add this custom type:

https://www.doctrine-project.org/projects/doctrine-dbal/en/2.9/reference/types.html#custom-mapping-types

Enum's aren't supported in DBAL, but for our purposes here, a custom class that is there just to add an additional type to DBAL so we can differentiate between `enum` and `string` is enough to do what we're looking for here.

If there's another way to do it, I'd be glad to hear about it. 😅

skeets23 commented:
You'll see the `CustomEnum` class is mentioned on line 76 of `GenerateCommand.php`

jasonmccreary commented on src/Types/EnumType.php:
```diff
@@ -0,0 +1,36 @@
+use Doctrine\DBAL\Types\Type;
+use Doctrine\DBAL\Platforms\AbstractPlatform;
+
+/**
```
Remove comments

jasonmccreary commented on src/Types/EnumType.php:
```diff
@@ -0,0 +1,36 @@
+class EnumType extends Type
+{
+
+    const ENUM = 'enum'; // modify to match your type name
```
Remove comment

jasonmccreary commented on src/Types/EnumType.php:
```diff
@@ -0,0 +1,36 @@
+
+    public function getSQLDeclaration(array $fieldDeclaration, AbstractPlatform $platform)
+    {
+        // this is wrong, but it also isn't important, since we're not modifying tables anyways. (this method must exist to match the abstract parent class, but we won't use it)
```
Improve comment - what _should_ this be under normal circumstances. Also wrap at 80 characters.

jasonmccreary commented on src/Console/GenerateCommand.php:
```diff
@@ -303,6 +306,7 @@ protected function setProperty($name, $type = null)
         }
 
         $fakeableTypes = [
+						'enum' => '$faker->randomElement(' . $this->enumValues($table, $name) . ')',
```
Fix indentation.

jasonmccreary commented on src/Console/GenerateCommand.php:
```diff
@@ -71,6 +73,7 @@ public function __construct(Filesystem $files, $view)
      */
     public function handle()
     {
+        Type::addType("customEnum", "Mpociot\LaravelTestFactoryHelper\Types\EnumType");
```
Using single quotes and leverage PHP `::class` reference.

skeets23 commented:
Thanks for the code review. It should be ready to go now. 👍

jasonmccreary commented:
Thanks. I'll merge this and tag it this week.
